numpy/core/src/multiarray/refcount.c

```diff
@@ -292,7 +292,8 @@ static void
 _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
 {
     if (!PyDataType_FLAGCHK(dtype, NPY_ITEM_REFCOUNT)) {
-        if ((obj == Py_None) || (PyInt_Check(obj) && PyLong_AsLong(obj)==0)) {
+        if ((obj == Py_None) ||
+                (PyLong_Check(obj) && PyLong_AsLong(obj) == 0)) {
```

Comment:
It looks to me like this can set an exception that doesn't get cleared or propagated, which can result in a systemerror.


Admittedly the old code was just as broken.

Reply:
I'll take another look.

Reply:
OK, I'm going to just clear any possible error, the function doesn't allow error handling. The whole thing probably needs to be redesigned.

Reply:
To be honest, I think you have to here. This function is mainly called during array deallocation and has to clear out references, so we could "print unraisable", but it should clear any error and continue work at the very least.

I suppose it might be nice if it did report errors (after finishing as well as possible?), even if the error ends up being ignored most of the time.